File: hw3/code/Problem2.m
%% Housekeeping
clear all
close all
clc

obstacle1 = {[3,0],[5,0],[3,2],[5,2]};
obstacle2 = {[6,-2],[8,-2],[6,0],[8,0]};
obstacles = {obstacle1,obstacle2};
%% i) Plot Vector Field _________________________________________________
figure
subplot(2,1,1)
x0=100;
y0=50;
width=800;
height=700;
set(gcf,'position',[x0,y0,width,height])
field_info = csvread('field_2a.csv');
meta_data = num2cell(field_info(1,1:5));
[top,bottom,left,buffer,scale_factor] = deal(meta_data{:});
mat_field = field_info(2:end,1:end-1);
[h,w] = size(mat_field);
for i=1:h
   for j=1:w/2
       quiver((left-buffer+j-1)/scale_factor,(top+buffer-i)/scale_factor,mat_field(i,2*(j-1)+1),mat_field(i,2*(j-1)+2),'b')
       hold on
   end
end
for obstacle = obstacles
   rectangle('Position',[obstacle{1}{1} abs(obstacle{1}{1}(1)-obstacle{1}{2}(1)) abs(obstacle{1}{2}(2)-obstacle{1}{3}(2))],'FaceColor',[0 .5 .5]);
end
grid on
grid minor
xlabel('q_1')
ylabel('q_2')
title('Vector Field')
%% ii)

fprintf("I started with a large dstar_goal and slowly decreased it \n")
fprintf("until I saw that the attractive forces were decreasing smoothly\n")
fprintf("as the robot reached the gap between the obstacles. The Qstar\n")
fprintf("for both obstacles were the same and the robot was able to move\n")
fprintf("straight towards the goal without deviating from the obstacles\n")
%% iii) Plot Path 

mat = csvread('path_2a.csv');

start = [0,0];
goal = [10,0];
subplot(2,1,2)
h = zeros(3,1);
h(1) = plot(start(1),start(2),'om','Linewidth',2);
hold on
h(3) = plot(goal(1),goal(2),'xk','Linewidth',2);
h(2) = plot((mat(:,1)-abs(left)-buffer)/scale_factor,(mat(:,2)-abs(bottom)-buffer)/scale_factor,'b','Linewidth',2);
for obstacle = obstacles
   rectangle('Position',[obstacle{1}{1} abs(obstacle{1}{1}(1)-obstacle{1}{2}(1)) abs(obstacle{1}{2}(2)-obstacle{1}{3}(2))]);
end
legend(h,'start','path','goal')
grid on
grid minor
%xlim([-5 5])
%ylim([-5 5])
xlabel('q_1')
ylabel('q_2')
title('Attractive-Repulsive Gradient Descent Path')

set(gcf,'position',[x0,y0,width,height])

%% iv)
len_path=0;
for i=1:length(mat)-1
    len_path = len_path + sqrt((mat(i+1,1)-mat(i,1))^2 + (mat(i+1,2)-mat(i,2))^2);
end
fprintf("The length of the path is: %f\n",len_path/scale_factor)

%% v)
fprintf("Yes, I would expect different path lengths since the Qstar \n")
fprintf("influences how close a robot can come to a certain obstacle\n")
fprintf("and since the obstacles are near the robots path to goal\n")
fprintf("the robot would surely be swayed if Qstar is decreased\n")

File: hw3/code/problem2.cpp
#include <iostream>
#include <cmath>
#include <vector>
#include <fstream>
#include "potentials.h"

using namespace std;

#define PI 3.14159265

int main(){
    Point qstart{0,0};
    Point qgoal{10,0};
    double alpha = 0.1;
    double zeta = 0.1;
    double dstar = 6;
    double eta = 1;
    vector<double> Qstar = {1,1};
    double epsilon = 0.25;
    Potentials planner(qstart, qgoal, alpha, zeta, epsilon, dstar, eta, Qstar);

    // 2 a)_______________________________________________________________________________
    // Retrieve obstacles
    int num_obstacles = 2;
    vector<Point> obstacle1{{3,2},{5,2},{3,0},{5,0}};
    vector<Point> obstacle2{{6,-2},{8,-2},{6,0},{8,0}};
    planner.retrieveObstacles(num_obstacles, obstacle1, obstacle2);//vector<Point> {{3,2},{5,2},{3,0},{5,0}},vector<Point> {{6,-2},{8,-2},{6,0},{8,0}});

    // Generate path for gradient descent algorithm
    planner.scale_factor = 4;
    planner.discretizeMap();
    vector<vector<Point>> field = planner.vectorField();

    
    // Write the vector field of the cspace to a csv file
    ofstream field_file;
    field_file.open ("field_2a.csv");
    field_file << planner.top << "," << planner.bottom << "," << planner.left << "," << planner.buffer << "," << planner.scale_factor << endl;
    for(int i=0; i<field.size(); i++){
        for(int j=0; j<field[0].size(); j++){
            field_file <<field[i][j].x << "," <<field[i][j].y << ",";
        }
        field_file << endl;
    }
    field_file.close();

    
    // Write the path of the robot to a csv file
    vector<Point> path = planner.gradientDescent();
    ofstream path_file;
    path_file.open ("path_2a.csv");
    for(int i=0; i<path.size(); i++){
        path_file << path[i].x << "," << path[i].y <<"\n";
    }
    path_file.close();

    
    
    return 0;
}